app/src/main/java/pm/gnosis/heimdall/ui/safe/pairing/PairingAuthenticator.kt

```diff
@@ -0,0 +1,184 @@
+    private val recoverSafeOwnersHelper: RecoverSafeOwnersHelper,
+    private val gnosisSafeRepository: GnosisSafeRepository,
+    private val tokenRepository: TokenRepository,
+    private val transactionExecutionRepository: TransactionExecutionRepository,
+    private val pushServiceRepository: PushServiceRepository,
+    private val moshi: Moshi
+) : PairingAuthenticatorContract() {
+
+    private val errorHandler = CoroutineExceptionHandler { _, e ->
+        viewModelScope.launch {
+            Timber.e(e)
+        }
+    }
+
+    override val observableState: LiveData<ViewUpdate>
+        get() = _state
+    private val _state = MutableLiveData<ViewUpdate>()
+
+    private lateinit var safeAddress: Solidity.Address
+
+    private lateinit var balanceChannel: ReceiveChannel<Pair<ERC20Token, BigInteger?>>
+
+    override fun setup(safeAddress: Solidity.Address) {
+        this.safeAddress = safeAddress
+    }
+
+    override fun estimate() {
+
+        viewModelScope.launch(Dispatchers.IO + errorHandler) {
+
+            val safeInfo = gnosisSafeRepository.loadInfo(safeAddress).awaitFirst()
+            val paymentToken = tokenRepository.loadPaymentToken(safeAddress).await()
+
+            val owner = safeInfo.owners[0]
+            val extension = Solidity.Address(BigInteger.valueOf(Long.MAX_VALUE))
+            val transaction = recoverSafeOwnersHelper.buildRecoverTransaction(
+                safeInfo,
+                safeInfo.owners.subList(2, safeInfo.owners.size).toSet(),
+                setOf(owner, extension)
+            )
+
+            balanceChannel = tokenRepository.loadTokenBalances(safeAddress, listOf(paymentToken))
+                .repeatWhen { it.delay(BALANCE_REQUEST_INTERVAL_SECONDS, TimeUnit.SECONDS) }
+                .retryWhen { it.delay(BALANCE_REQUEST_INTERVAL_SECONDS, TimeUnit.SECONDS) }
+                .map {
+                    it[0]
+                }
+                .openSubscription()
+
+
+            balanceChannel.consumeEach {
+
+                val balance = it.second ?: BigInteger.ZERO
+
+                val executeInfo = transactionExecutionRepository.loadExecuteInformation(safeAddress, paymentToken.address, transaction).await()
+
+                val gasFee = with(executeInfo) {
+                    (txGas + dataGas + operationalGas) * gasPrice
+                }
+                _state.postValue(
+                    ViewUpdate.Balance(
+                        false,
+                        paymentToken.displayString(balance),
+                        ERC20TokenWithBalance(paymentToken, gasFee).displayString(roundingMode = RoundingMode.UP),
+                        paymentToken.displayString(balance - gasFee),
+                        paymentToken.symbol,
+                        balance > gasFee
+
+                    )
+                )
+            }
+        }
+    }
+
+    override fun pair(payload: String) {
+
+        viewModelScope.launch(Dispatchers.IO + errorHandler) {
```

Comment:
we should probably use appDispatchers here

Reply:
yes